File: proyecto_madrina/normalizar.py
import re
from typing import Optional

from modelo import Bloque


_RE_HORA = re.compile(r"(\d{1,2})[:\.](\d{2})")


def texto(valor) -> str:
    if valor is None:
        return ""
    return str(valor).replace("\n", " ").strip()


def nombre_catedratico(valor) -> tuple[str, bool]:
    """Normaliza el nombre del catedrático y detecta el marcador (NUEVO)."""
    raw = texto(valor).upper()
    if not raw:
        return "", False
    es_nuevo = "(NUEVO)" in raw
    nombre = raw.replace("(NUEVO)", "").strip()
    nombre = re.sub(r"\s+", " ", nombre)
    return nombre, es_nuevo


def codigos(valor) -> tuple[str, list[str]]:
    """Una celda puede traer 'ICI3120 / ICI3117' o 'ADM2011 /ADM2015'. Devuelve principal + alternos."""
    raw = texto(valor)
    if not raw:
        return "", []
    partes = re.split(r"[\/,]", raw)
    limpios = [p.strip().upper() for p in partes if p.strip()]
    if not limpios:
        return "", []
    return limpios[0], limpios[1:]


def alumnos(valor) -> Optional[int]:
    """Suma '12 + 5' a 17. Devuelve None si no es interpretable."""
    if valor is None or valor == "":
        return None
    if isinstance(valor, (int, float)):
        return int(valor)
    raw = texto(valor).replace(" ", "")
    if not raw:
        return None
    numeros = re.findall(r"\d+", raw)
    if not numeros:
        return None
    return sum(int(n) for n in numeros)


def horas(valor) -> Optional[float]:
    if valor is None or valor == "":
        return None
    if isinstance(valor, (int, float)):
        return float(valor)
    raw = texto(valor)
    try:
        return float(raw)
    except ValueError:
        return None


def bloques_dia(dia: str, valor) -> list[Bloque]:
    """Una celda de día puede traer uno o más rangos: '7:00 - 8:50', '8:00-9:50 LAB',
    '13:00 - 14:50\\n15:00 - 16:50', '7:00 10:50' (sin guion), etc."""
    raw = texto(valor)
    if not raw:
        return []

    raw_limpio = re.sub(r"(?i)\b(virtual|lab|laboratorio|presencial)\b", " ", raw)

    horas_encontradas = _RE_HORA.findall(raw_limpio)
    if len(horas_encontradas) < 2:
        return []

    minutos = [int(h) * 60 + int(m) for h, m in horas_encontradas]

    bloques = []
    for i in range(0, len(minutos) - 1, 2):
        inicio, fin = minutos[i], minutos[i + 1]
        if fin <= inicio:
            continue
        bloques.append(Bloque(dia=dia, inicio=inicio, fin=fin))

    return bloques


def aula(valor) -> str:
    raw = texto(valor).upper()
    return re.sub(r"\s+", " ", raw)


def seccion(valor) -> str:
    raw = texto(valor).upper().replace(" ", "")
    return raw


def minutos_a_hhmm(m: int) -> str:
    return f"{m // 60:02d}:{m % 60:02d}"

File: web/lib/procesar/normalizar.ts
/**
 * Funciones de normalización — espejo de proyecto_madrina/normalizar.py.
 *
 * El Excel de programación trae los datos en formatos inconsistentes;
 * este módulo los lleva a estructuras canónicas.
 */
import type { Bloque } from "./modelo";
import type { Dia } from "@/lib/types";

const RE_HORA = /(\d{1,2})[:.](\d{2})/g;

export function texto(valor: unknown): string {
  if (valor === null || valor === undefined) return "";
  return String(valor).replace(/\n/g, " ").trim();
}

export function nombreCatedratico(valor: unknown): {
  nombre: string;
  esNuevo: boolean;
} {
  const raw = texto(valor).toUpperCase();
  if (!raw) return { nombre: "", esNuevo: false };
  const esNuevo = raw.includes("(NUEVO)");
  const limpio = raw.replace("(NUEVO)", "").replace(/\s+/g, " ").trim();
  return { nombre: limpio, esNuevo };
}

export function codigos(valor: unknown): {
  principal: string;
  alternos: string[];
} {
  const raw = texto(valor);
  if (!raw) return { principal: "", alternos: [] };
  const partes = raw
    .split(/[\/,]/)
    .map((p) => p.trim().toUpperCase())
    .filter(Boolean);
  if (partes.length === 0) return { principal: "", alternos: [] };
  return { principal: partes[0], alternos: partes.slice(1) };
}

export function alumnos(valor: unknown): number | null {
  if (valor === null || valor === undefined || valor === "") return null;
  if (typeof valor === "number") return Math.trunc(valor);
  const raw = texto(valor).replace(/\s/g, "");
  if (!raw) return null;
  const nums = raw.match(/\d+/g);
  if (!nums) return null;
  return nums.reduce((a, b) => a + parseInt(b, 10), 0);
}

export function horas(valor: unknown): number | null {
  if (valor === null || valor === undefined || valor === "") return null;
  if (typeof valor === "number") return valor;
  const raw = texto(valor);
  const n = Number(raw);
  return Number.isFinite(n) ? n : null;
}

export function bloquesDia(dia: Dia, valor: unknown): Bloque[] {
  const raw = texto(valor);
  if (!raw) return [];

  // Sacar anotaciones (LAB, VIRTUAL, etc.) que no son horas
  const limpio = raw.replace(/\b(virtual|lab|laboratorio|presencial)\b/gi, " ");

  const horasFound: number[] = [];
  let m;
  RE_HORA.lastIndex = 0;
  while ((m = RE_HORA.exec(limpio)) !== null) {
    horasFound.push(parseInt(m[1], 10) * 60 + parseInt(m[2], 10));
  }

  if (horasFound.length < 2) return [];

  const bloques: Bloque[] = [];
  for (let i = 0; i + 1 < horasFound.length; i += 2) {
    const inicio = horasFound[i];
    const fin = horasFound[i + 1];
    if (fin <= inicio) continue;
    bloques.push({ dia, inicio, fin });
  }
  return bloques;
}

export function aula(valor: unknown): string {
  return texto(valor).toUpperCase().replace(/\s+/g, " ");
}

export function seccion(valor: unknown): string {
  return texto(valor).toUpperCase().replace(/\s+/g, "");
}

export function campusDeSeccion(seccionNormalizada: string): "T" | "C" {
  const s = seccionNormalizada.toUpperCase().replace(/\s+/g, "");
  if (s.endsWith("-C") || s.endsWith("C")) return "C";
  return "T";
}

export function modalidadCanonica(valor: unknown): string {
  const m = texto(valor).toUpperCase().replace(/[\s-]/g, "");
  if (m.startsWith("PRESEN")) return "PRESENCIAL";
  if (m.startsWith("VIRT")) return "VIRTUAL";
  if (m.startsWith("SEMI")) return "SEMIPRESENCIAL";
  return texto(valor).toUpperCase();
}

export function minutosAHHMM(m: number): string {
  const h = Math.floor(m / 60).toString().padStart(2, "0");
  const mm = (m % 60).toString().padStart(2, "0");
  return `${h}:${mm}`;
}
